File: Objective-C/Objective-C/RPIView.m
//
//  RPIView.m
//  
//
//  Created by YoloHwang on 17/3/22.
//
//

#import "RPIView.h"

@implementation RPIView

#define defaultLineWidth   4.0
#define defaultDuration    5.0
#define defaultFillColor   [UIColor clearColor]
#define defaultStrokeColor [UIColor redColor]
#define RGB(r,g,b)         [UIColor colorWithRed:r/255.0 green:gr/255.0 blue:b/255.0 alpha:1.0]
#define defaultColors      [NSArray arrayWithObjects:(id)[UIColor redColor].CGColor, nil];
#define defaultLocations   [NSArray arrayWithObjects:@(1.0), nil];


- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.lineWidth   = defaultLineWidth;
        self.fillColor   = defaultFillColor;
        self.strokeColor = defaultStrokeColor;
        self.duration    = defaultDuration;
        self.colors      = defaultColors;
        self.locations   = defaultLocations;
    }
    return self;
}

- (void)showRoundProgressIndecatorWithParentView:(UIView *)parentView {
    
    [self setupUI];
    [parentView addSubview:self];
}

- (void)setupUI {
    
    float centerX = self.bounds.size.width * 0.5;
    float centerY = self.bounds.size.height * 0.5;
    float raduis = (self.bounds.size.width - self.lineWidth) * 0.5;
    
    UIBezierPath *path = [UIBezierPath bezierPathWithArcCenter:CGPointMake(centerX, centerY) radius:raduis startAngle:-M_PI_2 endAngle:(1.5 * M_PI) clockwise:YES];
    
    CAShapeLayer *shapeLayer = [CAShapeLayer layer];
    shapeLayer.fillColor = self.fillColor.CGColor;
    shapeLayer.strokeColor = self.strokeColor.CGColor;
    shapeLayer.lineWidth = self.lineWidth;
    shapeLayer.lineCap = kCALineCapRound;
    shapeLayer.path = path.CGPath;
    
    CAGradientLayer *gradientLayer = [CAGradientLayer layer];
    gradientLayer.frame = self.bounds;
    gradientLayer.colors = self.colors;
    gradientLayer.locations = self.locations;
    
    gradientLayer.startPoint = CGPointMake(0, 0);
    gradientLayer.endPoint = CGPointMake(0, 1);
    gradientLayer.mask = shapeLayer;
    
    [self.layer addSublayer:gradientLayer];
    [self animationWithLayer:shapeLayer];
}

- (void)animationWithLayer:(CALayer *)layer {
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    animation.fromValue = @(0);
    animation.toValue = @(1.0);
    animation.duration = self.duration;
    animation.fillMode = kCAFillModeForwards;
    animation.removedOnCompletion = NO;
    [layer addAnimation:animation forKey:nil];
}

@end

File: SwiftRPI/SwiftRPI/RPIView.swift
//
//  RPIView.swift
//  SwiftRPI
//
//  Created by YoloHwang on 17/3/28.
//  Copyright © 2017年 Yolo. All rights reserved.
//

import UIKit

class RPIView: UIView {
    
    public var lineWidth: CGFloat    = 4.0
    public var duration: CGFloat     = 5.0
    public var fillColor: UIColor    = UIColor.clear
    public var strokeColor: UIColor  = UIColor.red
    public var colors: [CGColor]     = [UIColor.red.cgColor,UIColor.blue.cgColor]
    public var locations: [NSNumber] = [0.3,0.6]
    
    public func showRoundProgressIndecatorWithParentView(_ parentView: UIView,hasGradient has: Bool ) {
        setupUI(has)
        parentView.addSubview(self)
    }
    
    fileprivate func setupUI(_ hasGradient: Bool) {
        let centerX = self.bounds.size.width * 0.5
        let centerY = self.bounds.size.height * 0.5
        let raduis = centerX - self.lineWidth * 0.5
        
        let path = UIBezierPath.init(arcCenter: CGPoint(x: centerX, y: centerY), radius: raduis, startAngle: CGFloat(-M_PI_2), endAngle: CGFloat(M_PI * 1.5), clockwise: true)
        
        let shapeLayer = CAShapeLayer()
        shapeLayer.path = path.cgPath
        shapeLayer.fillColor = self.fillColor.cgColor
        shapeLayer.strokeColor = self.strokeColor.cgColor
        shapeLayer.lineWidth = self.lineWidth
        shapeLayer.lineCap = kCALineCapRound
        shapeLayer.strokeEnd = 1.0
        
        if hasGradient {
            let gradientLayer = CAGradientLayer()
            gradientLayer.frame = self.bounds
            gradientLayer.colors = self.colors
            gradientLayer.locations = self.locations
            gradientLayer.startPoint = CGPoint (x: 0, y: 0)
            gradientLayer.endPoint = CGPoint (x: 1, y: 0)
            gradientLayer.mask = shapeLayer
            
            self.layer.addSublayer(gradientLayer)
        } else {
            self.layer.addSublayer(shapeLayer)
        }
        
        rpiAnimation(shapeLayer)
    }
    
    fileprivate func rpiAnimation(_ layer: CALayer) {
        let animation = CABasicAnimation (keyPath: "strokeEnd")
        animation.fromValue = 0
        animation.toValue = 1
        animation.duration = CFTimeInterval(self.duration)
        animation.fillMode = kCAFillModeForwards
        animation.isRemovedOnCompletion = false
        layer.add(animation, forKey: nil)
        
    }

}

